### PR TITLE
Feat: add the Tab chart and refactor render system 

### PR DESCRIPTION
**Pull request by chenjiandongx**
本次 PR 内容，

### CODE REVIEW

File: pyecharts/charts/base.py [hunk 1]
```diff
@@ -67,52 +59,20 @@ def render(
         env: Optional[Environment] = None,
         **kwargs,
     ) -> str:
-        self._prepare_render()
-        RenderEngine(env).render_chart_to_file(
-            template_name=template_name, chart=self, path=path, **kwargs
-        )
-        return os.path.abspath(path)
+        return super()._render(path, template_name, env, **kwargs)
```
Comment (chfw) [inline pyecharts/charts/base.py]: In this case super()._render(...) should be super().render(). No need to use a protected function.

default template name, 'simple_chart.html' could be an attribute of base.py

File: pyecharts/charts/base.py [hunk 2]
```diff
@@ -67,52 +59,20 @@ def render(
...
         template_name: str = "simple_chart.html",
         env: Optional[Environment] = None,
         **kwargs,
     ) -> str:
-        self._prepare_render()
-        html = RenderEngine(env).render_chart_to_template(
-            template_name, chart=self, **kwargs
-        )
-        return html
+        return super()._render_embed(template_name, env, **kwargs)
```
Comment (chfw) [inline pyecharts/charts/base.py]: same for _render_embed()

File: pyecharts/charts/composite_charts/page.py [hunk 3]
```diff
@@ -129,74 +124,30 @@ def _prepare_render(self):
...
         template_name: str = "simple_page.html",
         env: types.Optional[Environment] = None,
         **kwargs,
     ) -> str:
-        self._prepare_render()
-        RenderEngine(env).render_chart_to_file(
-            template_name=template_name, chart=self, path=path, **kwargs
-        )
-        return os.path.abspath(path)
+        return super()._render(path, template_name, env, **kwargs)
```
Comment (chfw) [inline pyecharts/charts/composite_charts/page.py]: when _render() become render(), this function is no longer required.

File: pyecharts/charts/composite_charts/page.py [hunk 4]
```diff
@@ -129,74 +124,30 @@ def _prepare_render(self):
...
         self,
         template_name: str = "simple_page.html",
         env: types.Optional[Environment] = None,
         **kwargs,
     ) -> str:
-        self._prepare_render()
-        return RenderEngine(env).render_chart_to_template(
-            template_name=template_name, chart=self, **kwargs
-        )
+        return super()._render_embed(template_name, env, **kwargs)
```
Comment (chfw) [inline pyecharts/charts/composite_charts/page.py]: same here.

File: pyecharts/charts/composite_charts/tab.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        return self
+
+    def __iter__(self):
+        for chart in self._charts:
+            yield chart
+
+    def __len__(self):
+        return len(self._charts)
+
+    def render(
```
Comment (chfw) [inline pyecharts/charts/composite_charts/tab.py]: useless render() if super._render() become public function

File: pyecharts/charts/composite_charts/tab.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        self,
+        path: str = "render.html",
+        template_name: str = "simple_tab.html",
+        env: types.Optional[Environment] = None,
+        **kwargs,
+    ) -> str:
+        return super()._render(path, template_name, env, **kwargs)
+
+    def render_embed(
```
Comment (chfw) [inline pyecharts/charts/composite_charts/tab.py]: same here

File: pyecharts/charts/composite_charts/tab.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        return super()._render_embed(template_name, env, **kwargs)
+
+    def _prepare_render(self):
+        for c in self:
+            c.json_contents = c.dump_options()
+            if c.theme not in ThemeType.BUILTIN_THEMES:
+                self.js_dependencies.add(c.theme)
+
+    def render_notebook(self):
+        for c in self:
```
Comment (chfw) [inline pyecharts/charts/composite_charts/tab.py]: wondering why this for-loop cannot be part of super()._render_notebook().

File: pyecharts/render/engine.py [hunk 8]
```diff
@@ -40,17 +48,92 @@ def render_chart_to_file(self, template_name: str, chart: Any, path: str, **kwar
...
+    def render(self):
+        raise NotImplementedError
+
+    def render_embed(self):
+        raise NotImplementedError
+
+    def render_notebook(self):
+        raise NotImplementedError
+
+    def _render(
```
Comment (chfw) [inline pyecharts/render/engine.py]: all three functions in this render class can be public

**Comment by chenjiandongx** [comment]
@chfw 

image

image

**Comment by chenjiandongx** [comment]
可以避免重名 因为签名不一定相同

File: pyecharts/charts/composite_charts/tab.py (hunk 7 above)
Comment (chenjiandongx) [reply pyecharts/charts/composite_charts/tab.py]: base 并不需要这个 for 循环

File: pyecharts/render/engine.py [hunk 9]
```diff
@@ -40,17 +48,92 @@ def render_chart_to_file(self, template_name: str, chart: Any, path: str, **kwar
...
+class Render:
+    def __int__(self):
+        self.js_host: str = ""
+        self.js_functions: utils.OrderedSet = utils.OrderedSet()
+        self.js_dependencies = utils.OrderedSet()
+
+    def _prepare_render(self):
+        raise NotImplementedError
+
+    def render(self):
```
Comment (chfw) [inline pyecharts/render/engine.py]: this render() is useless then

File: pyecharts/charts/base.py [hunk 10]
```diff
@@ -67,52 +59,20 @@ def render(
...
         **kwargs,
     ) -> str:
-        self._prepare_render()
-        RenderEngine(env).render_chart_to_file(
-            template_name=template_name, chart=self, path=path, **kwargs
-        )
-        return os.path.abspath(path)
+        return super()._render(path, template_name, env, **kwargs)
 
     def render_embed(
```
Comment (chfw) [inline pyecharts/charts/base.py]: python does not support method overloading

**Comment by chfw** [comment]
```
def render(*args, **kwds):
    pass

def render_embed(*args, **kwds):
    pass
```

File: pyecharts/render/engine.py (hunk 9 above)
Comment (chenjiandongx) [reply pyecharts/render/engine.py]: ok, will remove them

**Pull request merged**